### DISCUSSION TITLE
PipeTarget.ToDelegate overload

### DISCUSSION THREAD

**Discussion by delphianaus** [Tyrrrz/CliWrap#183] [Q&A]
is it possible to overload PipeTarget.ToDelegate so you can pass the commandEvent you normally get from ListenAsync() or add something like PipeTarget.ToListenAsync 

Also for the argments.add if you pass in a empty list could you skip adding "" if the list  is empty

this would be great

**Comment by Tyrrrz** [comment]
I mean, `.Observe()` is pretty much that. What exactly are you trying to achieve?

**Comment by delphianaus** [comment]
Yes thats it.


i am doing this for example

        var cmd = Cli.Wrap("ffmpeg.exe").
                        WithArguments(args => args
                       .Add(argsx)).
                       WithWorkingDirectory(defaultpath).ExecuteAsync();


I want to use observe in this above example. Some examples would be good.
things i want to observe are StartedCommandEvent, StandardErrorCommandEvent
& ExitedCommandEvent

Thanks great work


Justin

**Reply by Tyrrrz** [reply]
Do this:

```csharp
cmd.Observe().Subscribe(cmdEvent => /* pattern match */);
```

You might need to add the `System.Reactive.Linq` package

**Comment by delphianaus** [comment]
Can you do it before calling executeasync() ? like witharguments?

**Reply by Tyrrrz** [reply]
You don't need to call `ExecuteAsync()`. These models are mutually exclusive.

**Comment by delphianaus** [comment]
Hey i want to know how i can send a control-c to a process i have started
with Cli.Wrap

My Code is

var clidb1 = Cli.Wrap("mongod.exe").
                    WithArguments(args => args
                    .Add("--port").Add("40001")
                    .Add("--dbpath").Add($"C:/YouTubeHelper/Data/db1")

.Add("--replSet").Add("--YTReplSet").Add("--bind_ip_all")).

WithStandardErrorPipe(PipeTarget.ToDelegate(EDB1_Handler)).

WithStandardOutputPipe(PipeTarget.ToDelegate(DB1_Handler)).
                    ExecuteAsync().ConfigureAwait(false);

i need the process to end with control-c before cli.Wrap terminates is self
( ie application shutdown).

Kind Regards

Justin

**Reply by Tyrrrz** [reply]
https://github.com/Tyrrrz/CliWrap#timeout-and-cancellation

**Reply by aborn** [reply]
https://gist.github.com/ycherkes/b216e78ae7ee4418504320f861337ed4
This is the best method to stop process gracefully.
` var success = await WindowsProcessTerminator.StopProcessGracefully(process, cts.Token);`
Actually, it sends `Ctrl-C` to process.
Now, the problem is how to get Porcess obj. CliWrap provide processId。
`Process process = Process.GetProcessById(_processId);`

**Comment by delphianaus** [comment]
No Not That , i am talking about physically sending a control-c event to
the underlying process before killing it , the mongodb deamon on windows
will shutdown when you send a control c.


something like this i found worked for 4 process ( 3 for db ,1 for windows
service)

 Process[] ps12 = Process.GetProcessesByName("mongod");
                            foreach (var p in ps12)
                            {

GenerateConsoleCtrlEvent(ConsoleCtrlEvent.CTRL_C, (int)p.SessionId);
                                Thread.Sleep(150);
                                p.Kill();
                            }

 it would be handy if you could go send Key and also a Kill on the
cli.Wraps task it returns

Justin
